File: P8/DAA027.java
import java.util.Scanner;

public class DAA027 {
    static int nodes, edges;    // Número de nós e arestas
    static boolean m[][];       // Matriz de adjacências
    static boolean visited[];   // Visitados
    static int color[];         // Cores dos nós

    static final int GREEN = 0;
    static final int RED = 1;

    static boolean dfs(int v, int c) {
        if (visited[v]) {
            if (color[v] == c) return true;
            return false;
        } 
        color[v] = c;
        visited[v] = true;
        c = (c == GREEN) ? RED : GREEN;
        for(int i = 1; i <= nodes; i++) {
            if(m[v][i] && !dfs(i,c)) return false; 
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int casos = in.nextInt();           // Número de casos
        for (int c = 0; c < casos; c++) {
            nodes = in.nextInt();
            edges = in.nextInt();
            m = new boolean[nodes + 1][nodes + 1];
            visited = new boolean[nodes + 1];
            color = new int[nodes + 1];
            
            for(int i = 0; i <= nodes; i++) color[i] = -1;

            // Ligações
            for (int e = 0; e < edges; e++) {
                int a = in.nextInt();
                int b = in.nextInt();
                m[a][b] = m[b][a] = true;
            }
            
            if (dfs(1, GREEN))
                System.out.println("sim");
            else
                System.out.println("nao");
        }
        in.close();
    }
}

File: P8/DAA027.cpp
#include <iostream>

using namespace std;
#define MAX 55
#define GREEN 0
#define RED 1

int nodes;          // Número de nós
bool m[MAX][MAX];   // Matriz de adjacências
bool visited[MAX];  // Visitados
int color[MAX];     // Cores dos nós

void reset() {
    for(int i = 0; i < MAX; i++) {
        for(int j = 0; j < MAX; j++) 
            m[i][j] = false;
        visited[i] = false;
    }
}

bool dfs(int v, int c) {
    if(visited[v]) {
        if(color[v] == c) return true;
        return false;
    } else {
        color[v] = c;
        visited[v] = true;
        (c == GREEN) ? c = RED : c = GREEN;
        for (int i = 1; i <= nodes; i++) {
            if (m[v][i] && !dfs(i,c)) return false;
        }
        return true;
    }
}

int main() {
    int n;      // Número de casos a considerar
    scanf("%d", &n);
    for(int c = 0; c < n; c++) {
        int edges;
        scanf("%d %d", &nodes, &edges);
        
        for(int i = 0; i < MAX; i++) color[i] = -1;

        // Ligações
        int a, b;
        for(int e = 0; e < edges; e++) {
            scanf("%d %d", &a, &b);
            m[a][b] = m[b][a] = true;
        }
        
        if(dfs(1, GREEN)) printf("sim\n");
        else printf("nao\n");
        reset();
    }

    return 0;
}